services/gateway/gateway-service.mjs

```diff
@@ -178,15 +180,43 @@ class GatewayService {
    * - displayName: The name to show for this player.
    * - displayChar: The character asset to display.
    */
-  async onUserLocation(socket, msg) {
+    async onUserLocation(socket, msg) {
     msg.playerID = socket.playerID;
     msg.displayName = socket.playerData.displayName;
     msg.displayChar = socket.playerData.displayChar;
     // TODO: Check if facing is valid.
     // TODO: Check if movement is legal.
+    let mapShift = {x: 6 , y: 3};
+    let mapSize = this.gameMap.getMapSize();
+    let lastRecord = await this.broadcaster.gameState.getPlayer(socket.playerID);
+    // target position is in the map
+    if(!this._borderCheck(msg,mapSize,mapShift)){
+      // restrict user position
+      msg.x = lastRecord.x;
+      msg.y = lastRecord.y;
+      await this._broadcastUserLocation(msg);
+      return;
+    }
+
+    // near by grid check  
+    if(!this._nearByGridCheck(msg,lastRecord)){
+      return ;
+    }
+
+    // obstacle check
+    let players = this.broadcaster.gameState.getPlayers();
+    let playerIDs = Object.keys(players);
+    let obstacle = playerIDs.find( playerID => players[playerID].x === msg.x && players[playerID].y === msg.y)
```

Comment:
`players` might not have all the players because we can have multiple gateway servers.